### ISSUE TITLE
Firefox shows 'Fail: error' when logging in

### ISSUE THREAD

**Issue by ddeboer** [zotonic/zotonic#1877]
Firefox (Quantum 58.0.2) doesn’t seem to be happy about the WebSocket connection cleanup that was done in 1a2132654047fe50125bcab0d9f5822325357ad8 (introduced in Zotonic 0.35.0). While logging in continues to work, this is shown in the console:

![screen shot 2018-02-08 at 15 59 22](https://user-images.githubusercontent.com/89267/35979620-1014917e-0ce9-11e8-8078-d8667a2c947d.png)

Is this caused by a new session being started when the user logs in?

**Comment by mmzeeman** [comment]
What was the error firefox reported? 

I can't reproduce the error over here. Was there an ajax call when the user was logging out?

**Comment by ddeboer** [comment]
Have you tried Quantum with the dev console open? 

It seems to be come timing issue related to the session change (killing the page session processes and with that the WebSockets connection) when logging in.

**Comment by mmzeeman** [comment]
Yes I have.

Does the page have a websocket or a comet connection? Looking at the fix I guess it is a comet connection in this case.

**Comment by mmzeeman** [comment]
When I force the page to have a comet connection I see this message.

```ubf
Don't know where to delegate incoming message  #{'z_msg_v1',0,'false',32~dq13ZKxqymGRQzo8rOwb9zGwJsuXMwHT~,1518431445986,'ubf',7~session~,'page','undefined','undefined','undefined',{'auth_change',20~qmdtSD0lQHtvU9ovDl84~}}&$
```

**Comment by mmzeeman** [comment]
That doesn't look right. This message should have been handled.

**Comment by mmzeeman** [comment]
Interesting, this message is delivered as data without being parsed first.

**Comment by mmzeeman** [comment]
It looks like zotonic sends wrong data to the page.

When I put these two debug lines into z_transport_incoming_data

<img width="563" alt="screen shot 2018-02-12 at 12 25 43" src="https://user-images.githubusercontent.com/1024972/36094795-df2baf0c-0fef-11e8-8701-81e02aec8822.png">

I get this in the console log of the browser.

<img width="937" alt="screen shot 2018-02-12 at 12 26 43" src="https://user-images.githubusercontent.com/1024972/36094837-0c486e4e-0ff0-11e8-8aec-469c21967429.png">

Somehow an integer is added to the messages. The ubf parser skips the integer and returns the data with the integer removed. 

I smell something 🐠🐡🐟y

**Comment by mmzeeman** [comment]
The message is encoded twice...

**Comment by mmzeeman** [comment]
@mworrell What I see that the message encoded twice here: https://github.com/zotonic/zotonic/blob/0.x/modules/mod_base/controllers/controller_comet.erl#L76

This is what I get from a ?DEBUG trace here.

```
2018-02-12 12:02:04.920 [info] <0.13501.47> DEBUG: controller_comet:78  {
{z_msg_v1,0,false,<<"$comet-VJRYM99PNglH8CgHbheM-1">>,1518436910678,ubf,'$comet',undefined,desktop,undefined,<<"VJRYM99PNglH8CgHbheM">>,1},
<<"#{'z_msg_v1',0,'false',32~nEynt5siwPWnY5UQ0hnrhqiZmrpBwEJT~,1518436924920,'ubf',7~session~,'page','undefined','undefined','undefined',{'auth_change',20~VJRYM99PNglH8CgHbheM~}}&$">>}
```

I'm not sure if `z_transport:incoming` should already encode the data and that `z_transport:data_encode` is not needed here.

After `z_transport:data_encode` I get:

```
<<"177~#{'z_msg_v1',0,'false',32~nEynt5siwPWnY5UQ0hnrhqiZmrpBwEJT~,1518436924920,'ubf',7~session~,'page','undefined','undefined','undefined',{'auth_change',20~VJRYM99PNglH8CgHbheM~}}&$~$">>
```

As final result which is sent to the page. This is double encoded ubf.

**Comment by mmzeeman** [comment]
@ddeboer 

https://github.com/zotonic/zotonic/blob/0.x/src/support/z_session_page.erl#L346

I think this is the **real** problem. On an auth change, the websocket directly gets an encoded ubf message.

But when there is a comet process attached to the page it gets an ubf encoded response. This response is later ubf encoded again. The page will then get a response which it can't handle.

Could you check if this fix in `z_page_session.erl` also gets rid of the firefox issue you see?

```erlang
%% The user of the session changed, signal any connected push connections, stop this page.
handle_cast(auth_change, #page_state{comet_pid=CometPid, websocket_pid=WsPid} = State) ->
    Msg = z_transport:msg(page, <<"session">>, #auth_change{page_id=State#page_state.page_id}, []),
    case WsPid of
        undefined when is_pid(CometPid) ->
            CometPid ! {final, [Msg]};
        undefined ->
            ok;
        _ ->
            {ok, Data} = z_ubf:encode([Msg]),
            controller_websocket:websocket_send_data(WsPid, Data)
    end,
    {stop, normal, State};
```

**Comment by mmzeeman** [comment]
@ddeboer I pushed the patch to your fix branch.

**Comment by mworrell** [comment]
Fixed in #1878
